### PR TITLE
Fixes possible `Index was outside the bounds of the array` /VS + MSBuild

### LINKED ISSUE #8810: 'Index was outside the bounds of the array' when GeneratePackageOnBuild=true
MSBuild 16.3.2+e481bbf88

```xml
...
GenerateNuspec:
  Successfully created package 'D:\tmp\_Issues\VS2019\NugetErrIndex\ClassLibrary1\ClassLibrary1\bin\Debug\ClassLibrary1.1.0.0.nupkg'.
C:\Program Files\dotnet\sdk\3.0.100\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): error : Index was outside the bounds of the array. [D:\tmp\_Issues\VS2019\NugetErrIndex\ClassLibrary1\ClassLibrary1\ClassLibrary1.csproj]
Done Building Project "D:\tmp\_Issues\VS2019\NugetErrIndex\ClassLibrary1\ClassLibrary1\ClassLibrary1.csproj" (default targets) -- FAILED.
...
```

VisualStudio.16.Release/16.3.8+29503.13

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		Index was outside the bounds of the array.	ClassLibrary1	C:\Program Files\dotnet\sdk\3.0.100\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets	198	
```

How to reproduce:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
  </PropertyGroup>
  
  <ItemGroup>
    <!-- add any file, its name start with 'build', for example: -->
    <None Include="buildfoo.bar" Pack="true" PackagePath="" />
  </ItemGroup>

</Project>
```

### PR DESCRIPTION
**Pull request by 3F**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8810

## Testing/Validation

Tests Added: Yes 
Validation: steps to reproduce from https://github.com/NuGet/Home/issues/8810 and this:

```xml
<NuGetPackTaskAssemblyFile>D:\tmp\_Issues\VS2019\NugetErrIndex\PR\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\debug\netstandard2.0\ilmerge\NuGet.Build.Tasks.Pack.dll</NuGetPackTaskAssemblyFile>
```

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs [hunk 1]
```diff
@@ -38,7 +38,7 @@ public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
             var warnings = new List<PackagingLogMessage>();
             foreach (var folder in _folders)
             {
-                var files = builder.GetFiles().Where(t => t.StartsWith(folder));
+                var files = builder.GetFiles().Where(t => t.StartsWith(folder + Path.DirectorySeparatorChar));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]: The path separators should be normalized like: https://github.com/NuGet/NuGet.Client/blob/48f67796ff38e98e3dfe4ab55fce000577b74fea/src/NuGet.Core/NuGet.Packaging/Rules/ContentFolderInPackageReferenceProjectRule.cs#L24-L26

**Review by nkolev92 — COMMENTED** [review]
Thanks for the contribution! This looks like a great start. 
One more addition and some tests would be great!

Can you please add a test here; https://github.com/NuGet/NuGet.Client/blob/50cd6de4c3f1b284f71a7f23ed1e7c714fbe5ff0/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs#L1152

and 

https://github.com/NuGet/NuGet.Client/blob/48f67796ff38e98e3dfe4ab55fce000577b74fea/test/NuGet.Core.Tests/NuGet.Packaging.Test/UpholdBuildConventionRuleTests.cs#L57

**Comment by 3F** [comment]
I updated tests. Also added processing of `PathUtility.GetPathWithDirectorySeparator` together with `Path.DirectorySeparatorChar` use.


About *UpholdBuildConventionRuleTests:* Please clarify context because this will be failed anyway:

https://github.com/NuGet/NuGet.Client/blob/48f67796ff38e98e3dfe4ab55fce000577b74fea/test/NuGet.Core.Tests/NuGet.Packaging.Test/UpholdBuildConventionRuleTests.cs#L57

I mean, if I will test it like "buildfoo.bar", "build.log", ... and so on.

**Because** of this call:

https://github.com/NuGet/NuGet.Client/blob/48f67796ff38e98e3dfe4ab55fce000577b74fea/test/NuGet.Core.Tests/NuGet.Packaging.Test/UpholdBuildConventionRuleTests.cs#L28

And this **hard coded** logic with folders:

https://github.com/NuGet/NuGet.Client/blob/50cd6de4c3f1b284f71a7f23ed1e7c714fbe5ff0/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs#L70-L72

v

https://github.com/NuGet/NuGet.Client/blob/50cd6de4c3f1b284f71a7f23ed1e7c714fbe5ff0/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs#L95-L97


So let me know, does this related to other errors in your code or this is expected behaviour.

If it's error, then I can add something like this:

```csharp
if(filePath.IndexOf('/') == -1) {
    return "/";
}
```
with updated related tests

File: src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs (hunk 1 above)
Comment (3F) [reply src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]: done

**Comment by kartheekp-ms** [comment]
@3F - Thanks for your contribution. Sorry for the delay in responding to this pull request. I will be actively working on this PR going forward. Can you please `rebase` your private fork with the latest `dev` branch? I will review the proposed changes in the meantime.

**Comment by 3F** [comment]
Rebased onto dac23f49d 

@kartheekp-ms,

As I remember there was something wrong with UpholdBuildConventionRule and/or its tests. Please [review carefully](https://github.com/NuGet/NuGet.Client/pull/3118#issuecomment-553530547) before merge

File: src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs [hunk 2]
```diff
@@ -29,7 +29,10 @@ public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
             var warnings = new List<PackagingLogMessage>();
             foreach (var folder in _folders)
             {
-                var files = builder.GetFiles().Where(t => t.StartsWith(folder));
+                var files = builder.GetFiles()
+                                    .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                                    .Where(t => t.StartsWith(folder + Path.DirectorySeparatorChar));
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]:
```cs 
using System.IO;
```

Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]:
```suggestion
                                    .Where(t => t.StartsWith(folder + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase));
```

File: src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs [hunk 3]
```diff
@@ -44,7 +47,14 @@ internal IEnumerable<PackagingLogMessage> GenerateWarnings(IEnumerable<Conventio
             foreach (var folder in _folders)
             {
                 var warningMessage = new StringBuilder();
-                var currentConventionViolators = conventionViolators.Where(t => t.ExpectedPath.StartsWith(folder + '/'));
+                var currentConventionViolators = conventionViolators
+                                                    .Where(t =>
+                                                        PathUtility.GetPathWithDirectorySeparator(
+                                                            t.ExpectedPath
+                                                        )
+                                                        .StartsWith(folder + Path.DirectorySeparatorChar)
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]:
```suggestion
                                                        .StartsWith(folder + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
```

**Review by kartheekp-ms — CHANGES_REQUESTED** [review]
I have not yet completed reviewing this PR but I left few comments.

File: src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs (hunk 2 above)
Comment (3F) [reply src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]: Oh, I did not compile it again after your request for rebasing :)

**Comment by 3F** [comment]
@kartheekp-ms 

> I have not yet completed reviewing this PR but I left few comments.

Just to be sure, this PR focused on fixing problem https://github.com/NuGet/Home/issues/8810. However, [your requested suggestions](https://github.com/NuGet/NuGet.Client/pull/3118/files/bedee109aa8906ad351dcac526e9600a6fdc1715#r399041083) on the contrary *extends* possible collection of the convention violators and pkg validation. 

I don't know more about specification for this cases in nuget project, but I think we need add either new PR for requested changes or maybe just to split in a separate commits. Let me know details.

**Comment by kartheekp-ms** [comment]
> Just to be sure, this PR focused on fixing problem [NuGet/Home#8810](https://github.com/NuGet/Home/issues/8810). However, [your requested suggestions](https://github.com/NuGet/NuGet.Client/pull/3118/files/bedee109aa8906ad351dcac526e9600a6fdc1715#r399041083) on the contrary _extends_ possible collection of the convention violators and pkg validation.
> 
> I don't know more about specification for this cases in nuget project, but I think we need add either new PR for requested changes or maybe just to split in a separate commits. Let me know details.

I think it is okay to consider [above suggestions](https://github.com/NuGet/NuGet.Client/pull/3118#discussion_r399040932) in this PR because we are modifying the same lines of code to fix NuGet/Home#8810. Similar code review comment was provided [earlier](https://github.com/NuGet/NuGet.Client/pull/3118#discussion_r345520461). We have good test coverage for this scenario so we can identify issues before merging. Let us know If you still think otherwise so that I can create new issue to invoke overloaded method that accepts  `StringComparison.OrdinalIgnoreCase`.

**Comment by 3F** [comment]
No problem, as you are prefer. Updated + added new tests

> because we are modifying the same lines of code to fix

:) The question was about various logic for convention violators and pkg validation *compared to the original*.

> Similar code review comment was provided earlier

Well, one more unified abstraction is needed if all this are the same case. Just to avoid misunderstanding.

File: src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs (hunk 2 above)
Comment (3F) [reply src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs]: Updated with new added tests

**Comment by kartheekp-ms** [comment]
> Well, one more unified abstraction is needed if all this are the same case. Just to avoid misunderstanding

Can you please give an example so that I can understand the context better?

> The question was about various logic for convention violators and pkg validation compared to the original.

Upon further review, I think we should also modify below methods. We don't need to normalize paths in convention violators collection now because the paths in source collection are normalized.

https://github.com/NuGet/NuGet.Client/blob/69276891c4a2e391b3711c6b7d5276886a5d857b/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs#L86

- The new implementation for `GetFolderName` method may look like below

```cs
        internal string GetFolderName(string filePath)
        {
            string[] paths = filePath.Split(Path.DirectorySeparatorChar);
            string folder = paths[0];

            NuGetFramework framework = NuGetFramework.ParseFolder(paths[1]);
            if (framework != NuGetFramework.UnsupportedFramework)
            {
                folder = Path.Combine(folder, framework.GetShortFolderName());
            }

            return PathUtility.EnsureTrailingSlash(folder);
        }
```
- Change the first argument from `group.First()` to `group.Key`
https://github.com/NuGet/NuGet.Client/blob/69276891c4a2e391b3711c6b7d5276886a5d857b/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs#L78

- `Path = filePath` is sufficient because we care about parent of `file`
https://github.com/NuGet/NuGet.Client/blob/69276891c4a2e391b3711c6b7d5276886a5d857b/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs#L106

The above modifications doesn't fix this issue because almost all the `tests` have hardcoded forward slash `/` char in the path which results in test failures. For e.g.,
https://github.com/NuGet/NuGet.Client/blob/731d470b14c9ff94558aa06156a50e7a0e272b88/test/NuGet.Core.Tests/NuGet.Packaging.Test/UpholdBuildConventionRuleTests.cs#L94

`UpholdBuildConventionRuleTests.cs` has so many occurrences of forward slash `/` char hardcoded in the path which results in test failures. I will discuss internally to decide the best way to fix tests.

Please feel free to let us know if you have time to contribute further or else I will take up this task and complete the pending work.

**Comment by 3F** [comment]
> Can you please give an example so that I can understand the context better?

I meant at least something like *t => t.PathStartsWith(...)* that will aggregate a common case in all Packaging/Rules/* for an *.StartsWith(...+ Path.DirectorySeparatorChar, ...*

Or better, new additional object that will provide a common extractor for such as a convention violators objects or others that will implement specific interface.

> The new implementation for GetFolderName method may look like below
> ...
> The above modifications doesn't fix this issue because almost all the tests have hardcoded forward slash / char

I recommend also temp throwing an exception in this temp solution where *paths[1]* still is blurred for any purposes.

> I will discuss internally to decide the best way to fix tests.

Good.

> Please feel free to let us know if you have time to contribute further or else I will take up this task and complete the pending work.

It actually depends on the scale of this task.  And as I see, here are big problems with hardcoded logic and lot of dependencies on it. 

Thus, more like i'm not ready to continue review of additional parts where something like *('/')[0] + '/' + hi + '/'* for the whole NuGet client today. Sorry.

**Comment by kartheekp-ms** [comment]
@3F - @zivkan is working on fixing a duplicate issue in this [PR](https://github.com/NuGet/NuGet.Client/pull/3353). Hence closing this PR. Thank you for this contribution.

**Pull request closed without merging**